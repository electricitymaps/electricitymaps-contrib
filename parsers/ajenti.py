"""
Initial PR https://github.com/electricitymaps/electricitymaps-contrib/pull/2456
Discussion thread https://github.com/electricitymaps/electricitymaps-contrib/issues/636
A promotion webpage for King's Island energy production is here: https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
As of 09/2020, it embeds with <iframe> the URI https://data.ajenti.com.au/KIREIP/index.html

About the data, the feed we get seems to be counters with a 2 seconds interval.
That means that if we fetch these counters every 15 minutes, we only are reading "instantaneous" metters that could differ from the total quantity of energies at play.
To get the very exact data, we would need to have a parser running constanty to collect those 2-sec interval counters.
"""

import json
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session
from signalr import Connection

ZONE_PARAMS = {
    # King Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
    "AU-TAS-KI": {
        "hub": "TagHub",
        "method": "Dashboard",
        "tz": "Australia/Currie",
        "source": "hydro.com.au",  # Iframe: https://data.ajenti.com.au/KIREIP/index.html
    },
    # Flinders Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2533
    # https://en.wikipedia.org/wiki/Flinders_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/flinders-island
    "AU-TAS-FI": {
        "hub": "flindershub",
        "method": "SendDashboard",
        "tz": "Australia/Hobart",
        "source": "hydro.com.au",
    },
    # Rottnest Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2534
    # https://en.wikipedia.org/wiki/Rottnest_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/rottnest-island
    "AU-WA-RI": {
        "hub": "HogsHub",
        "method": "SendDashboard",
        "tz": "Australia/Perth",
        "source": "hydro.com.au",
    },
}


class SignalR:
    def __init__(self, url):
        self.url = url

    def update_res(self, msg):
        if msg != {}:
            self.res = msg

    def get_value(self, hub, method):
        self.res = {}
        with Session() as session:
            # create a connection
            connection = Connection(self.url, session)
            chat = connection.register_hub(hub)
            chat.client.on(method, self.update_res)
            connection.start()
            connection.wait(3)
            connection.close()
            return self.res


def parse_payload(logger: Logger, payload) -> dict:
    technologies_parsed = {
        "biomass": 0,
        "battery": 0,
        "coal": 0,
        "flywheel": 0,
        "gas": 0,
        "hydro": 0,
        "nuclear": 0,
        "oil": 0,
        "solar": 0,
        "wind": 0,
        "geothermal": 0,
        "unknown": 0,
    }
    if "technologies" not in payload:
        raise KeyError(
            f"No 'technologies' in payload\n" f"serie : {json.dumps(payload)}"
        )
    else:
        logger.debug(f"serie : {json.dumps(payload)}")
    for technology in payload["technologies"]:
        # rename upstream key to match our payload needs
        if technology["id"] == "diesel":
            technology["id"] = "oil"

        assert technology["unit"] == "kW"
        # The upstream API gives us kW, we need MW
        technologies_parsed[technology["id"]] = int(technology["value"]) / 1000

    # King's Island uses some percentage of biodiesel; since we account differently traditionnal oil & biomass, here we do math to separate both
    if "biodiesel" in payload:
        # sanity check to be sure that we are not remaking data that the data source could otherwise give us
        assert "biomass" and "oil" != 0
        technologies_parsed["biomass"] = (
            technologies_parsed["oil"] * payload["biodiesel"]["percent"] / 100
        )
        technologies_parsed["oil"] = (
            technologies_parsed["oil"] * (100 - payload["biodiesel"]["percent"]) / 100
        )

    logger.debug(f"production : {json.dumps(technologies_parsed)}")

    return technologies_parsed


# Both keys battery and flywheel are negative when storing energy, and positive when feeding energy to the grid
def format_storage_techs(technologies_parsed):
    # sometimes we don't get those keys. For instance Rottnest island don't have any storage.
    if "battery" and "flywheel" in technologies_parsed:
        storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]
    else:
        storage_techs = 0
    _battery_production = (
        storage_techs if storage_techs > 0 else 0
    )  # TODO: Should this just be removed?
    _battery_storage = storage_techs if storage_techs < 0 else 0


def sum_storage_techs(technologies_parsed):
    storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]

    return storage_techs


def fetch_production(
    zone_key: str = "AU-TAS-KI",
    session: Session | None = None,
    target_datetime=None,
    logger: Logger = getLogger(__name__),
) -> dict:
    if target_datetime is not None:
        raise NotImplementedError(
            "The datasource currently implemented is only real time"
        )

    # get the specific zone parameters
    try:
        hub, dashboard, tz, source = (
            ZONE_PARAMS[zone_key]["hub"],
            ZONE_PARAMS[zone_key]["method"],
            ZONE_PARAMS[zone_key]["tz"],
            ZONE_PARAMS[zone_key]["source"],
        )
    except KeyError as e:
        raise KeyError("The zone " + zone_key + " isn't implemented") from e

    payload = SignalR("https://data.ajenti.com.au/live/signalr").get_value(
        hub, dashboard
    )
    technologies_parsed = parse_payload(logger, payload)
    storage_techs = sum_storage_techs(technologies_parsed)

    return {
        "zoneKey": zone_key,
        "datetime": datetime.now(tz=ZoneInfo(tz)),
        "production": {
            "biomass": technologies_parsed["biomass"],
            "coal": technologies_parsed["coal"],
            "gas": technologies_parsed["gas"],
            "hydro": technologies_parsed["hydro"],
            "nuclear": technologies_parsed["nuclear"],
            "oil": technologies_parsed["oil"],
            "solar": technologies_parsed["solar"],
            "wind": 0
            if technologies_parsed["wind"] < 0 and technologies_parsed["wind"] > -0.1
            else technologies_parsed[
                "wind"
            ],  # If wind between 0 and -0.1 set to 0 to ignore self-consumption
            "geothermal": technologies_parsed["geothermal"],
            "unknown": technologies_parsed["unknown"],
        },
        "storage": {
            "battery": storage_techs
            * -1  # Somewhat counterintuitively,to ElectricityMap positive means charging and negative means discharging
        },
        "source": source,
    }


if __name__ == "__main__":
    print(fetch_production())
