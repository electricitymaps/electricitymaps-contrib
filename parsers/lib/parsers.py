import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
