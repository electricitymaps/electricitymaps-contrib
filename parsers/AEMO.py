import io
import re
import zipfile
from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import TotalConsumptionList
from electricitymap.contrib.lib.models.events import EventSourceType
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "aemo.com.au"

ZONE_KEY_TO_REGION = {
    "AU-NSW": "NSW1",
    "AU-QLD": "QLD1",
    "AU-SA": "SA1",
    "AU-TAS": "TAS1",
    "AU-VIC": "VIC1",
    "AU-WA": "WEM",  # This zone is not implemented yet
}

ZONE_KEY_TO_TIMEZONE = {
    "AU-NSW": ZoneInfo("Australia/Sydney"),
    "AU-QLD": ZoneInfo("Australia/Brisbane"),
    "AU-SA": ZoneInfo("Australia/Adelaide"),
    "AU-TAS": ZoneInfo("Australia/Hobart"),
    "AU-VIC": ZoneInfo("Australia/Melbourne"),
    "AU-WA": ZoneInfo("Australia/Perth"),  # This zone is not implemented yet
}

# TODO, what about the other zone in Australia (AU-WA)? Check remaining zone


def find_document(session, target_datetime):
    # Fetch the directory listing
    base_url = "http://nemweb.com.au/Reports/CURRENT/Operational_Demand/FORECAST_HH/"
    response = session.get(base_url)

    # Parse with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")

    # String datetime
    target_date_str = target_datetime.strftime("%Y%m%d")

    # Find matching links
    matching_links = soup.find_all(
        "a",
        href=re.compile(
            rf"PUBLIC_FORECAST_OPERATIONAL_DEMAND_HH_{target_date_str}\d+_\d+\.zip"
        ),
    )

    if matching_links:
        target_link = matching_links[-1]  # Get the last (most recent) link
        full_url = urljoin(base_url, target_link["href"])  # Construct full URL

        file_response = session.get(full_url)

        with zipfile.ZipFile(io.BytesIO(file_response.content)) as z:
            csv_filename = z.namelist()[0]
            with z.open(csv_filename) as f:
                df = pd.read_csv(f)
                return df
    else:
        print("No matching files found")


def fetch_consumption_forecast(
    zone_key: ZoneKey,  # "AU-NSW", "AU-QLD", "AU-SA", "AU-TAS", "AU-VIC", "AU-WA"
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Consumption forecast in MW every half an hour for 10 days ahead.
    Only for NSW1, QND1, SA1, TAS1, VIC1 zones."""
    session = session or Session()

    if target_datetime is None:
        target_datetime = datetime.now(tz=ZONE_KEY_TO_TIMEZONE[zone_key])

    df = find_document(session, target_datetime)

    # Transform dataframe
    df.columns = df.iloc[0]
    df = df.iloc[1:-1].reset_index(drop=True)

    #
    region = ZONE_KEY_TO_REGION.get(zone_key)
    all_consumption_events = df[
        df["REGIONID"] == region
    ]  # all events with a datetime and a consumption value
    consumption_list = TotalConsumptionList(logger)
    for _, event in all_consumption_events.iterrows():
        datetime_object = datetime.strptime(
            event["INTERVAL_DATETIME"], "%Y/%m/%d %H:%M:%S"
        ).replace(tzinfo=ZONE_KEY_TO_TIMEZONE[zone_key])

        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime_object,
            consumption=float(
                event["OPERATIONAL_DEMAND_POE50"]
            ),  # 50% probability of exceedance operational demand forecast value
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the electricityMap backend, but handy for testing."""

    print(fetch_consumption_forecast("AU-NSW"))
    print(fetch_consumption_forecast("AU-QLD"))
    print(fetch_consumption_forecast("AU-SA"))
    print(fetch_consumption_forecast("AU-TAS"))
    print(fetch_consumption_forecast("AU-VIC"))
    print(
        fetch_consumption_forecast("AU-WA")
    )  # Not implemented yet. It returns an empty list
