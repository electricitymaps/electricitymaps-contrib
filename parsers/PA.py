import json
import re
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from parsers.lib.exceptions import ParserException

PARSER = "PA.py"
TIMEZONE = ZoneInfo("America/Panama")
ZONE_KEY = ZoneKey("PA")

CONSUMPTION_URL = "https://sitr.cnd.com.pa/m/pub/sin.html"
CONSUMPTION_SOURCE = "sitr.cnd.com.pa"

EXCHANGE_URL = "https://sitr.cnd.com.pa/m/pub/int.html"
EXCHANGE_SOURCE = "sitr.cnd.com.pa"

PRODUCTION_URL = "https://sitr.cnd.com.pa/m/pub/gen.html"
PRODUCTION_SOURCE = "sitr.cnd.com.pa"

# Sources:
# 1. https://www.celsia.com/Portals/0/contenidos-celsia/accionistas-e-inversionistas/perfil-corporativo-US/presentaciones-US/2014/presentacion-morgan-ingles-v2.pdf
# 2. https://www.celsia.com/en/about-celsia/business-model/power-generation/thermoelectric-power-plants
# 3. https://endcoal.org/tracker/
# 4. http://aesmcac.com/aespanamades/en/colon/ "It reuses the heat from the exhaust gas from the gas turbines in order to obtain steam, to be later used by a steam turbine and to save fuel consumption in the production of electricity."
# 5. https://panamcham.com/sites/default/files/el_inicio_del_futuro_del_gas_natural_en_panama.pdf "3 gas turbines and 1 steam (3X1 configuration)" "Technology: Combined Cycle" | This and the previous source taken together seems to imply that the steam turbine is responsible for the second cycle of the CCGT plant, giving confidence that output from all four units should indeed be tallied under "gas". Furthermore, as the plant also has a LNG import facility it is most unlikely the steam turbine would be burning a different fuel such as coal or oil.
# 6. https://www.etesa.com.pa/documentos/Tomo_II__Plan_Indicativo_de_Generacin_2019__2033.pdf page 142
# 7. http://168.77.210.79/energia/wp-content/uploads/sites/2/2020/08/2-CEE-1970-2019-GE-Generaci%C3%B3n-El%C3%A9ctrica.xls (via http://www.energia.gob.pa/mercado-energetico/?tag=84#documents-list)
# 8. https://www.asep.gob.pa/wp-content/uploads/electricidad/resoluciones/anno_12528_elec.pdf
# 9. https://www.irena.org/-/media/Files/IRENA/Agency/Publication/2018/May/IRENA_RRA_Panama_2018_En.pdf
# 10. https://www.asep.gob.pa/wp-content/uploads/electricidad/concesiones_licencias/concesiones_licencias/2021/listado_licencias_abr27.pdf
# 11. https://generadoragatun.com/
MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE = {
    "ACP Miraflores 2": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "ACP Miraflores 5": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "ACP Miraflores 6": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "ACP Miraflores 7": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "ACP Miraflores 8": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "ACP Miraflores 9": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "ACP Miraflores 10": "oil",  # [7] Sheet "C-GE-1A-1 CapInstXEmp"
    "BLM 2": "coal",  # [7] Sheet "C-GE-1A-2 CapInstXEmp"
    "BLM 3": "coal",  # [7] Sheet "C-GE-1A-2 CapInstXEmp"
    "BLM 4": "coal",  # [7] Sheet "C-GE-1A-2 CapInstXEmp"
    "BLM 5": "oil",  # [7] Sheet "C-GE-1A-2 CapInstXEmp"
    "BLM 6": "oil",  # [7] Sheet "C-GE-1A-2 CapInstXEmp"
    "BLM 8": "oil",  # [7] Sheet "C-GE-1A-2 CapInstXEmp"
    "BLM 9": "oil",  # [7] Sheet "C-GE-1A-2 CapInstXEmp" mentions no fuel type, and given all other units are accounted for this must be the heat recovery boiler for the 3 diesel-fired units mentioned in [2]
    "CADASA 1": "biomass",  # [9]
    "CADASA 2": "biomass",  # [9]
    "Cativá 1": "oil",  # [1][2]
    "Cativá 2": "oil",  # [1][2]
    "Cativá 3": "oil",  # [1][2]
    "Cativá 4": "oil",  # [1][2]
    "Cativá 5": "oil",  # [1][2]
    "Cativá 6": "oil",  # [1][2]
    "Cativá 7": "oil",  # [1][2]
    "Cativá 8": "oil",  # [1][2]
    "Cativá 9": "oil",  # [1][2]
    "Cativá 10": "oil",  # [1][2]
    "Cobre Panamá 1": "coal",  # [3]
    "Cobre Panamá 2": "coal",  # [3]
    "Costa Norte 1": "gas",  # [4][5]
    "Costa Norte 2": "gas",  # [4][5]
    "Costa Norte 3": "gas",  # [4][5]
    "Costa Norte 4": "gas",  # [4][5]
    "Esperanza 1": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Esperanza 2": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Esperanza 3": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Esperanza 4": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Esperanza 5": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Esperanza 6": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Esperanza 7": "oil",  # [7] has a single 92MW bunker fuel power plant, but [8] shows this is actually a power barge with 7 units
    "Jinro": "oil",  # [6][7]
    "Pacora 1": "oil",  # [6]
    "Pacora 2": "oil",  # [6]
    "Pacora 3": "oil",  # [6]
    "PanAm 1": "oil",  # [6][7]
    "PanAm 2": "oil",  # [6][7]
    "PanAm 3": "oil",  # [6][7]
    "PanAm 4": "oil",  # [6][7]
    "PanAm 5": "oil",  # [6][7]
    "PanAm 6": "oil",  # [6][7]
    "PanAm 7": "oil",  # [6][7]
    "PanAm 8": "oil",  # [6][7]
    "PanAm 9": "oil",  # [6][7]
    "Proyecto Gatún 1": "gas",  # [11]
    "Proyecto Gatún 2": "gas",  # [11]
    "Proyecto Gatún 3": "gas",  # [11]
    "Sparkle Power 1": "oil",  # [10]
    "Sparkle Power 2": "oil",  # [10]
    "Sparkle Power 3": "oil",  # [10]
    "Sparkle Power 4": "oil",  # [10]
    "Sparkle Power 5": "oil",  # [10]
    "Sparkle Power 6": "oil",  # [10]
    "Sparkle Power 7": "oil",  # [10]
    "Sparkle Power 8": "oil",  # [10]
    "Termocolón 1": "oil",  # [6] (spelled "Termo Colón")
    "Termocolón 2": "oil",  # [6] (spelled "Termo Colón")
    "Termocolón 3": "oil",  # [6] (spelled "Termo Colón")
    "Tropitérmica 1": "oil",  # [6]:162[7] spelled "Tropitermica" in both
    "Tropitérmica 2": "oil",  # [6]:162[7] spelled "Tropitermica" in both
    "Tropitérmica 3": "oil",  # [6]:162[7] spelled "Tropitermica" in both
}

PRODUCTION_TYPE_TO_PRODUCTION_MODE = {
    "Hídrica": "hydro",
    "Eólica": "wind",
    "Solar": "solar",
    "Biogás": "biomass",
    "Térmica": "unknown",
}

_SPANISH_CALENDAR = {
    "enero": "01",
    "febrero": "02",
    "marzo": "03",
    "abril": "04",
    "mayo": "05",
    "junio": "06",
    "julio": "07",
    "agosto": "08",
    "septiembre": "09",
    "octubre": "10",
    "noviembre": "11",
    "diciembre": "12",
}


def _localise_spanish_date(date: str) -> str:
    """Localises a date containing full name (lowercase) spanish months, replacing them with zero-padded decimal numbers.

    This avoids having to mess up the global locale to be able to parse the date.
    """
    return re.sub(
        "|".join(_SPANISH_CALENDAR.keys()),
        lambda m: _SPANISH_CALENDAR[m.group(0)],
        date,
    )


def extract_pie_chart_data(html):
    """Extracts generation breakdown pie chart data from the source code of the page"""
    # Extract object with data
    data_source = re.search(r"var localPie = (\[\{.+\}\]);", html).group(1)
    # Un-quoted keys ({key:"value"}) are valid JavaScript but not valid JSON (which requires {"key":"value"}).
    # Will break if other keys than these three are introduced.
    # Alternatively, use a JSON5 library (JSON5 allows un-quoted keys)
    data_source = re.sub(r"(name|value|color)", r'"\1"', data_source)
    return json.loads(data_source)


def sum_thermal_units(soup) -> float:
    """Sums thermal units of the generation mix to prevent using slightly outdated chart data.

    Thermal total from the graph and the total one would get from summing output of all generators deviates a bit,
    presumably because they aren't updated at the exact same moment.
    """

    # Sum thermal units from table Térmicas (MW)
    thermal_h3 = soup.find("h3", string=re.compile(r"\s*Térmicas\s*"))
    thermal_tables = thermal_h3.find_next_sibling().find_all(
        "table", {"class": "table table-hover table-striped table-sm sitr-gen-group"}
    )

    thermal_units = 0
    for thermal_table in thermal_tables:
        thermal_units += sum(
            [
                float(span.text)
                for span in thermal_table.find_all("span", {"style": "color:#222"})
            ]
        )
        thermal_units += sum(
            [
                float(span.text)
                for span in thermal_table.find_all("span", {"style": "color:ROYALBLUE"})
            ]
        )

    return thermal_units


def fetch_production(
    zone_key: ZoneKey = ZONE_KEY,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""

    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    # Fetch page and load into BeautifulSoup
    r = session or Session()
    response = r.get(PRODUCTION_URL)
    if not response.ok:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    response.encoding = "utf-8"
    html_doc = response.text
    soup = BeautifulSoup(html_doc, "html.parser")

    # Parse the datetime and return a python datetime object
    spanish_date = soup.find("h3", {"class": "sitr-update"}).string
    english_date = _localise_spanish_date(spanish_date)
    date = datetime.strptime(english_date, "%d-%m-%Y %H:%M:%S").replace(tzinfo=TIMEZONE)

    production_mix = ProductionMix()
    productions = extract_pie_chart_data(html_doc)
    # [{name:"Hídrica 1342.54 (80.14%)",value:1342.54,color:"#99ccee"}, ...]
    for prod in productions:  # {name:"Hídrica 1342.54 (80.14%)", ...}
        production_type, production_value, _percentage = prod["name"].split(" ", 2)
        # ignore termica data to avoid using outdated chart data
        if production_type == "Térmica":
            continue
        production_mode = PRODUCTION_TYPE_TO_PRODUCTION_MODE[production_type]  # hydro
        production_mix.add_value(production_mode, float(production_value))

    # Calculate manually thermal generation to avoid using outdated chart data
    thermal_generation = sum_thermal_units(soup)
    # Known fossil plants: parse, subtract from "unknown", add to "coal"/"oil"/"gas"
    thermal_production_breakdown = soup.find_all("table", {"class": "sitr-table-gen"})[
        1
    ]
    # Make sure the table header is indeed "Térmicas (MW)" (in case the tables are re-arranged)
    thermal_production_breakdown_table_header = (
        thermal_production_breakdown.parent.parent.parent.select("> .tile-title")[
            0
        ].string
    )
    assert "Térmicas" in thermal_production_breakdown_table_header, (
        f"Exception when extracting thermal generation breakdown for {zone_key}: table header does not contain "
        f"'Térmicas' but is instead named {thermal_production_breakdown_table_header}"
    )
    thermal_production_units = thermal_production_breakdown.select(
        "tbody tr td table.sitr-gen-group tr"
    )

    for thermal_production_unit in thermal_production_units:
        unit_name_and_generation = thermal_production_unit.find_all("td")
        unit_name = unit_name_and_generation[0].string
        unit_generation = float(unit_name_and_generation[1].string)
        if unit_name in MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE:
            if unit_generation > 0:  # Ignore self-consumption
                unit_fuel_type = MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE[
                    unit_name
                ]
                production_mix.add_value(unit_fuel_type, unit_generation)
                thermal_generation -= unit_generation
        else:
            logger.warning(
                f"{unit_name} is not mapped to generation type",
                extra={"key": zone_key},
            )

    # Round remaining "unknown" output to 13 decimal places to get rid of floating point errors
    thermal_generation = round(thermal_generation, 13)
    if 0 < thermal_generation < 1e-3:
        thermal_generation = 0.0
    # assign anything remaining to 'unknown'
    production_mix.add_value("unknown", max(thermal_generation, 0.0))

    production_breakdown_list = ProductionBreakdownList(logger)
    production_breakdown_list.append(
        zoneKey=zone_key,
        datetime=date,
        source=PRODUCTION_SOURCE,
        production=production_mix,
    )
    return production_breakdown_list.to_list()


def fetch_exchange(
    zone_key1: ZoneKey = ZONE_KEY,
    zone_key2: ZoneKey = ZoneKey("CR"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known power exchange (in MW) between two countries."""

    sorted_zone_keys = ZoneKey("->".join(sorted([zone_key1, zone_key2])))

    if target_datetime is not None:
        raise ParserException(
            PARSER,
            "This parser is not yet able to parse historical data",
            sorted_zone_keys,
        )

    session = session or Session()
    timestamp = datetime.now(tz=TIMEZONE)
    response = session.get(EXCHANGE_URL)
    if not response.ok:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            sorted_zone_keys,
        )

    df = pd.read_html(response.text)[0]

    # A positive value on website indicates a flow from country specified to PA.
    net_flow_cr = round(
        float(df[4][1])
        + float(df[4][3])
        + float(df[4][5])
        + float(df[1][8])
        + float(df[1][10]),
        2,
    )
    net_flow_gt = round(
        float(df[4][23]) + float(df[4][26]) + float(df[4][28]) + float(df[1][31]), 2
    )
    net_flow_hn = round(
        float(df[1][13])
        + float(df[1][15])
        + float(df[1][18])
        + float(df[1][20])
        + float(df[1][23]),
        2,
    )
    net_flow_ni = round(
        float(df[4][8]) + float(df[4][10]) + float(df[4][13]) + float(df[4][15]), 2
    )

    # invert sign to account for direction in alphabetical order
    net_flow_sv = -1 * round(
        float(df[4][18]) + float(df[4][20]) + float(df[1][26]) + float(df[1][28]), 2
    )

    net_flows = {
        "CR->PA": net_flow_cr,  # Costa Rica to Panama
        "GT->PA": net_flow_gt,  # Guatemala to Panama
        "HN->PA": net_flow_hn,  # Honduras to Panama
        "NI->PA": net_flow_ni,  # Nicaragua to Panama
        "PA->SV": net_flow_sv,  # Panama to El Salvador
    }

    exchange_list = ExchangeList(logger)
    exchange_list.append(
        zoneKey=sorted_zone_keys,
        datetime=timestamp,
        netFlow=net_flows[sorted_zone_keys],
        source=EXCHANGE_SOURCE,
    )
    return exchange_list.to_list()


def fetch_consumption(
    zone_key: ZoneKey = ZONE_KEY,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetches consumption of Panama."""

    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    r = session or Session()
    timestamp = datetime.now(tz=TIMEZONE)
    response = r.get(CONSUMPTION_URL)
    if not response.ok:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    soup = BeautifulSoup(response.text, "html.parser")
    consumption_title = soup.find("h5", string=re.compile(r"\s*Demanda Total\s*"))
    consumption_val = float(consumption_title.find_next_sibling().text.split()[0])

    consumption_list = TotalConsumptionList(logger)
    consumption_list.append(
        zoneKey=zone_key,
        datetime=timestamp,
        consumption=consumption_val,
        source=PRODUCTION_SOURCE,
    )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_exchange() ->")
    print(fetch_exchange())
    print("fetch_consumption() ->")
    print(fetch_consumption())
