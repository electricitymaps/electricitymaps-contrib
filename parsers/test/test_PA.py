from importlib import resources

import pytest
from freezegun import freeze_time
from requests_mock import ANY, GET

from parsers.PA import fetch_production


@pytest.fixture(autouse=True)
def mock_response(adapter):
    adapter.register_uri(
        GET,
        ANY,
        text=resources.files("parsers.test.mocks")
        .joinpath("PA_nominal_generation.html")
        .read_text(),
        status_code=200,
    )


@freeze_time("2021-12-30 09:58:40", tz_offset=-5)
def test_fetch_production(session, snapshot):
    assert snapshot == fetch_production(session=session)
