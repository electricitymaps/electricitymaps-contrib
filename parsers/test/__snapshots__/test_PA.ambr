# serializer version: 1
# name: test_fetch_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': FakeDatetime(2021, 12, 30, 9, 58, 37, tzinfo=zoneinfo.ZoneInfo(key='America/Panama')),
      'production': dict({
        'biomass': 2.75,
        'coal': 149.6,
        'gas': 355.88,
        'hydro': 421.84,
        'oil': 238.2,
        'solar': 262.76,
        'unknown': 0.0,
        'wind': 115.4,
      }),
      'source': 'sitr.cnd.com.pa',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'PA',
    }),
  ])
# ---
